src.related/pbmplus/ppm/ppmtoicr.1
.TH ppmtoicr 1 "30 July 1990"
.SH NAME
ppmtoicr - convert a portable pixmap into NCSA ICR format 
.SH SYNOPSIS
.B ppmtoicr
.RB [ -windowname
.IR name ]
.RB [ -expand
.IR expand ]
.RB [ -display
.IR display ]
.RB [ -rle ]
.RI [ ppmfile ]
.SH DESCRIPTION
Reads a portable pixmap file as input.
Produces an NCSA Telnet Interactive Color Raster graphic file as output.
If
.I ppmfile
is not supplied, 
.I ppmtoicr
will read from standard input.
.PP
Interactive Color Raster (ICR) is a protocol for displaying raster
graphics on workstation screens. The protocol is implemented in NCSA
Telnet for the Macintosh version 2.3. The ICR protocol shares
characteristics of the Tektronix graphics terminal emulation protocol.
For example, escape sequences are used to control the display.
.PP
.I ppmtoicr
will output the appropriate sequences to create a window of the
dimensions of the input pixmap,
create a colormap of up to 256
colors on the display, then load the picture data into the window.
.PP
Note that there is no icrtoppm tool - this transformation is one way.
.SH OPTIONS
.TP 14
.BI -windowname " name"
Output will be displayed in
.I name
(Default is to use
.I ppmfile
or "untitled" if standard input is read.)
.TP
.BI -expand " expand"
Output will be expanded on display by factor 
.I expand
(For example, a value of 2 will cause four pixels to be displayed for
every input pixel.)
.TP
.BI -display " display"
Output will be displayed on screen numbered 
.I display
.TP
.B -rle
Use run-length encoded format for display. (This will nearly always
result in a quicker display, but may skew the colormap.)
.SH EXAMPLES
To display a
.I ppm
file using the protocol:
.nf
    ppmtoicr ppmfile
.fi
This will create a window named 
.I ppmfile
on the display with the correct dimensions for
.I ppmfile,
create and download a colormap of up
to 256 colors, and download the picture into the window. The same effect
may be achieved by the following sequence:
.nf
    ppmtoicr ppmfile > filename
    cat filename
.fi
To display a GIF 
file using the protocol in a window titled after the input file, zoom
the displayed image by a factor of 2, and
run-length encode the data:
.nf
    giftoppm giffile | ppmtoicr -w giffile -r -e 2
.fi
.I ppmtoicr
will also read
.I pbm
and
.I pgm
files. The full range of pbm manipulation routines may be applied to
graphics for display using
.I ppmtoicr.
Thus, a monochrome Sun raster file may be viewed in 5 color
grayscale by the following sequence:
.nf
    pnmtorast rasterfile | pnmscale .5 | ppmtoicr
.fi
.SH BUGS
.PP
The protocol uses frequent 
.I fflush
calls to speed up display. If the
output is saved to a file for later display via
.I cat,
drawing will be
much slower. In either case, increasing the Blocksize limit on the
display will speed up transmission substantially.
.SH SEE ALSO
.BR ppm(5)
.LP
.IR "NCSA Telnet for the Macintosh", 
University of Illinois at Urbana-Champaign (1989)
.SH AUTHOR
Copyright (C) 1990 by Kanthan Pillay (svpillay@Princeton.EDU),
Princeton University Computing and Information Technology.
.\" Permission to use, copy, modify, and distribute this software and its
.\" documentation for any purpose and without fee is hereby granted, provided
.\" that the above copyright notice appear in all copies and that both that
.\" copyright notice and this permission notice appear in supporting
.\" documentation.  This software is provided "as is" without express or
.\" implied warranty.


CHANGES
- GISLIB 
  - G_warning(), G_fatal_error() - send mail only if eviroment variable
                                   GRASS_ERROR_MAIL is set
  - G_debug() - new, print debugging info on selected level
  - G_plot_icon() - new, plot selected mark
- DBMI 
  - dbmscap file is not used, drivers are searched in dist.$ARCH/driver/db
  - new dbf driver
- vector library completely rewritten (multilayer, multiformat, 3D)
- new make system


src.related/pbmplus/ppm/ppmtoicr.c
/* ppmtoicr.c - convert a portable pixmap to NCSA ICR protocol
**
** Copyright (C) 1990 by Kanthan Pillay (svpillay@Princeton.EDU)
**
** Permission to use, copy, modify, and distribute this software and its
** documentation for any purpose and without fee is hereby granted, provided
** that the above copyright notice appear in all copies and that both that
** copyright notice and this permission notice appear in supporting
** documentation.  This software is provided "as is" without express or
** implied warranty.
*/

#include "ppm.h"
#include "ppmcmap.h"

#define MAXCOLORS 256
#define CLUTCOLORS 768

pixel **pixels;
colorhash_table cht;
char *testimage;

void main(argc, argv)
int argc;
char *argv[];
{
	FILE *ifp;
	int argn, rows, cols, colors, i, j, BitsPerPixel, newxsize;
	pixval maxval;
	colorhist_vector chv;
	char rgb[CLUTCOLORS];
	int GetPixel();
	char *windowname, *thischar, *thisline, *space;
	register unsigned char c;
	register char *p;
	int display, expand;
	int rleflag, winflag;
	char *usage = "[-windowname windowname] [-expand expand] [-display display] [-rle] [ppmfile]";

	ppm_init( &argc, argv );

	argn = 1;
	windowname = "untitled";
	winflag = 0;
	expand = 1;
	display = 0;
	rleflag = 0;

	while ( argn < argc && argv[argn][0] == '-' && argv[argn][1] != '\0' )
	    {
	    if ( pm_keymatch(argv[argn],"-windowname",2) && argn + 1 < argc )
		{
		++argn;
		windowname = argv[argn];
		winflag = 1;
		}
	    else if ( pm_keymatch(argv[argn],"-expand",2) && argn + 1 < argc )
		{
		++argn;
		if ( sscanf( argv[argn], "%d",&expand ) != 1 )
		    pm_usage( usage );
		}
	    else if ( pm_keymatch(argv[argn],"-display",2) && argn + 1 < argc )
		{
		++argn;
		if ( sscanf( argv[argn], "%d",&display ) != 1 )
		    pm_usage( usage );
		}
	    else if ( pm_keymatch(argv[argn],"-rle",2) )
		rleflag = 1;
	    else if ( pm_keymatch(argv[argn],"-norle",2) )
		rleflag = 0;
	    else
		pm_usage( usage );
	    }

	if ( argn < argc )
	    {
	    ifp = pm_openr( argv[argn] );
	    if ( ! winflag )
		windowname = argv[argn];
	    ++argn;
	    }
	else
	    ifp = stdin;

	if ( argn != argc )
	    pm_usage( usage );

	pixels = ppm_readppm( ifp, &cols, &rows, &maxval );

	pm_close( ifp );

	for (i = 0; i < CLUTCOLORS; i++)
	    rgb[i] = 0;

	/* Figure out the colormap. */
	pm_message("computing colormap...", 0,0,0,0,0);
	chv = ppm_computecolorhist(pixels, cols, rows, MAXCOLORS, &colors);
	if (chv == (colorhist_vector) 0)
	pm_error(
	       "too many colors - try doing a 'ppmquant %d'", MAXCOLORS,
	       0,0,0,0);
	pm_message("%d colors found", colors, 0,0,0,0);

	/* Turn the ppm colormap into an ICR colormap. */
	if (maxval > 255)
	pm_message(
		"maxval is not 255 - automatically rescaling colors", 0,0,0,0,0);
	for (i = 0; i < colors; i++)
	{
	j = (3 * i);
	if (maxval == 255)
		{
		rgb[j] = PPM_GETR(chv[i].color) ;
		j++;
		rgb[j] = PPM_GETG(chv[i].color) ;
		j++;
		rgb[j] = PPM_GETB(chv[i].color) ;
		}
	else
		{
		rgb[j] = (int) PPM_GETR(chv[i].color) * 255 / maxval;
		j++;
		rgb[j] = (int) PPM_GETG(chv[i].color) * 255 / maxval;
		j++;
		rgb[j] = (int) PPM_GETB(chv[i].color) * 255 / maxval;
		}
	}
	BitsPerPixel = colorstobpp(colors);

	/* And make a hash table for fast lookup. */
	cht = ppm_colorhisttocolorhash(chv, colors);
	ppm_freecolorhist(chv);


	/************** Create a new window using ICR protocol *********/
	/* Format is "ESC^W;left;top;width;height;display;windowname"  */

	pm_message("creating window %s ...", windowname, 0,0,0,0);
	(void)printf("\033^W;%d;%d;%d;%d;%d;%s^",0,0,cols*expand,rows*expand,display,windowname);
	fflush(stdout);


	/****************** Download the colormap.  ********************/
	pm_message("downloading colormap for %s ...", windowname, 0,0,0,0);

	(void)printf("\033^M;%d;%d;%d;%s^",0,MAXCOLORS,CLUTCOLORS,windowname);
	thischar = rgb;
	for (j=0; j<CLUTCOLORS; j++) {
	c = *thischar++;
		if (c > 31 && c < 123 ) {	 /* printable ASCII */
		putchar(c);
		}
		else {
		putchar((c>>6)+123);	 /* non-printable, so encode it */
		putchar((c & 0x3f) + 32);
		}
	}
	fflush(stdout);

	/**************** send out picture *************************/
	/* Protocol's RLE scheme is quicker but buggy              */

	if (rleflag) {	
		pm_message("sending run-length encoded picture data ...", 0,0,0,0,0);
		testimage = (char*) malloc(rows*cols);
		p = testimage;
		for (i=0; i<rows; i++)
			for (j=0; j<cols; j++) 
			*p++ = GetPixel(j,i);
		space = (char*) malloc(rows*3);
		thisline = testimage;
		for (i = 0; i < rows; i++) {
			newxsize = rleit(thisline,space,cols);
			thisline += cols;	/* increment to next line */
		(void)printf("\033^R;%d;%d;%d;%d;%s^",0,i*expand,expand,newxsize,windowname);
		thischar = space;
		for (j=0; j< newxsize; j++) {
			c= *thischar++;  /*get byte to send */
			if (c>31 && c <123) {
				putchar(c);
				}
			else {
				putchar((c>>6) + 123);
				putchar((c & 0x3f) + 32);
				}
			}
			fflush(stdout);
		}
		free(space);
		exit(0);
		}

	/* Otherwise, send out uncompressed pixel data via the slow method */

		else {
		pm_message("sending picture data ...", 0,0,0,0,0);
		for (i = 0; i < rows; i++) {
			(void)printf("\033^P;%d;%d;%d;%d;%s^",0,i*expand,expand,cols,windowname);
			for (j = 0; j < cols; j++) {
				c  = GetPixel(j,i);
				if (c > 31 && c < 123) {
						putchar(c);
						}
				else		{
						putchar((c>>6)+123);
						putchar((c & 0x3f) + 32);
						}
				}
			}
		fflush(stdout);
		exit(0);
		}
	}

colorstobpp(colors)
int colors;
	{
	int bpp;

	if (colors <= 2)
	bpp = 1;
	else if (colors <= 4)
	bpp = 2;
	else if (colors <= 8)
	bpp = 3;
	else if (colors <= 16)
	bpp = 4;
	else if (colors <= 32)
	bpp = 5;
	else if (colors <= 64)
	bpp = 6;
	else if (colors <= 128)
	bpp = 7;
	else if (colors <= 256)
	bpp = 8;
	else
	pm_error("can't happen", 0,0,0,0,0);
	return bpp;
	}

GetPixel(x, y)
int x, y;
	{
	int color;

	color = ppm_lookupcolor(cht, &pixels[y][x]);
	return color;
	}


/* rleit   compress with run length encoding as per NCSA's documentation */

rleit(buf,bufto,len)
	int len;
	char *buf,*bufto;
	{
	register char *p,*q,*cfoll,*clead;
	char *begp;
	int i;

	p = buf;
	cfoll = bufto;
	clead = cfoll + 1;

	begp = p;
	while (len > 0 ) {		/* encode until gone */
		
		q = p + 1;
		i = len-1;
	while (*p == *q && i+120 > len && i) {
		q++;
		i--;
	}

	if (q > p +2) {			/* three in a row */
		if (p > begp) {
			*cfoll = p - begp;
			cfoll = clead;
		}
		*cfoll++ = 128 | (q-p);		/*len of seq*/
		*cfoll++ = *p;			/* char of seq */
		len -= q-p;		/* subtract len of seq */
		p = q;
		clead = cfoll+1;
		begp = p;
	}
	else {
		*clead++ = *p++;	/* copy one char */
		len--;
		if (p>begp + 120) {
			*cfoll = p - begp;
			cfoll = clead++;
			begp = p;
		}
	}
	}

/* fillin last bytecount */

	if (p>begp)
		*cfoll = (p - begp);
	else
		clead--;

	return((int) (clead-bufto));	/*how many stored as encoded */
}
